File: src/test/kotlin/info/davidvedvick/seis739/catpics/users/given an existing user without a password.kt
package info.davidvedvick.seis739.catpics.users

import info.davidvedvick.seis739.catpics.security.BadCatEmployeeCredentials
import info.davidvedvick.seis739.catpics.security.CatEmployeeCredentials
import info.davidvedvick.seis739.catpics.security.CatEmployeeEntry
import info.davidvedvick.seis739.catpics.security.UnauthenticatedCatEmployee
import io.mockk.coEvery
import io.mockk.mockk
import kotlinx.coroutines.runBlocking
import org.amshove.kluent.`should be instance of`
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class `given an existing user without a password` {
    @Nested
    inner class `when logging the user in` {
        private val services by lazy {
            CatEmployeeEntry(
                mockk {
                    coEvery { findByEmail("ZtyPVt") } returns CatEmployee(315, "ZtyPVt", "", true)
                },
                mockk(),
            )
        }

        private lateinit var catEmployeeCredentials: CatEmployeeCredentials

        @BeforeAll
        fun act() {
            runBlocking {
                catEmployeeCredentials = services.authenticate(UnauthenticatedCatEmployee("ZtyPVt", "MnI875"))
            }
        }

        @Test
        fun `then the user is not authenticated`() {
            catEmployeeCredentials `should be instance of` BadCatEmployeeCredentials::class
        }
    }
}

File: server/users/specs/given an existing user without a password.ts
import {beforeAll, describe, expect, test} from "@jest/globals";
import CatEmployeeEntry from "../CatEmployeeEntry.js";
import CatEmployee from "../CatEmployee.js";
import {ManageCatEmployees} from "../ManageCatEmployees.js";
import Encoder from "../../security/Encoder.js";
import BadCatEmployeeCredentials from "../BadCatEmployeeCredentials.js";

describe("given an existing user without a password", () => {
    describe("when logging the user in", () => {
        let error: BadCatEmployeeCredentials;

        beforeAll(async () => {
            const catEmployeeEntry = new CatEmployeeEntry({
                findByEmail(email: string): Promise<CatEmployee | null> {
                    return Promise.resolve(email == "ZtyPVt" ? {
                        email: email,
                        password: "",
                        isEnabled: true,
                        id: 315,
                    } : null);
                }
            } as ManageCatEmployees, {} as Encoder);

            try {
                await catEmployeeEntry.authenticate({
                    email: "ZtyPVt",
                    password: "MnI875",
                });
            } catch (e) {
                if (e instanceof BadCatEmployeeCredentials)
                    error = e;
            }
        });

        test("then the email is correct", () => {
            expect(error.email).toBe("ZtyPVt");
        });

        test("then the password is correct", () => {
            expect(error.password).toBe("MnI875");
        });
    });
});